File: new-resource-detector/src/main/java/nl/haploid/octowight/service/ResourceDetectorsService.java
package nl.haploid.octowight.service;

import nl.haploid.octowight.AtomChangeEvent;
import nl.haploid.octowight.AtomGroup;
import nl.haploid.octowight.detector.ResourceDetector;
import nl.haploid.octowight.registry.data.ResourceRoot;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.Collection;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class ResourceDetectorsService { // TODO: rename?

	@Autowired
	private List<ResourceDetector> detectors;

	protected List<ResourceDetector> getDetectors() {
		return detectors;
	}

	public List<ResourceDetector> getDetectorsForAtomType(final String atomType) {
		return this.getDetectors().stream()
				.filter(d -> d.getAtomTypes().contains(atomType))
				.collect(Collectors.toList());
	}

	public List<ResourceRoot> detectResources(final AtomGroup atomGroup, final List<AtomChangeEvent> events) {
		return getDetectorsForAtomType(atomGroup.getAtomType()).stream()
				.map(d -> d.detect(events))
				.flatMap(Collection::stream)
				.collect(Collectors.toList());
	}
}

File: new-resource-detector/src/main/scala/nl/haploid/octowight/service/ResourceDetectorsService.scala
package nl.haploid.octowight.service

import java.util

import nl.haploid.octowight.detector.ResourceDetector
import nl.haploid.octowight.{AtomChangeEvent, AtomGroup}
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

import scala.collection.JavaConverters._

@Service
class ResourceDetectorsService {
  @Autowired private[this] val detectors: util.List[ResourceDetector] = null

  def getDetectors: List[ResourceDetector] = detectors.asScala.toList

  def getDetectorsForAtomCategory(atomCategory: String) = getDetectors.filter(_.getAtomCategories.contains(atomCategory))

  def detectResources(atomGroup: AtomGroup, events: Traversable[AtomChangeEvent]) = {
    getDetectorsForAtomCategory(atomGroup.getCategory).flatMap(_.detect(events))
  }
}
